### PR TITLE
Fix ambiguity of erie token location after tile lay

### LINKED ISSUE #801: 1836Jr: Allow NBDS to move it's token after upgrading from a tile without edges to a tile with edges

### PR DESCRIPTION
**Pull request by jenf**
fixes #801

on NBDS it will automatically lay the token and sort the ambiguity out after the tile is layed
existing 1836jr games will likely be broken again

### CODE REVIEW

File: assets/app/view/game/hex.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def render
           when :track, :token_or_track
             opaque = @round.connected_hexes[@hex]
             clickable ||= opaque
-          when :token, :route, :home_token
+          when :token, :route, :home_token, :move_token_due_to_ambigious_city
```
Comment (tobymao) [inline assets/app/view/game/hex.rb:47]: can we change the name to be :reposition_token or :move_token

File: assets/app/view/game/part/city_slot.rb [hunk 2]
```diff
@@ -52,7 +52,18 @@ def on_click(event)
 
           # If there's a choice of tokens of different types show the selector, otherwise just place
           next_tokens = @game.current_entity.tokens_by_type
-          if next_tokens.size == 1 || @game.round.step == :home_token
+          if @game.round.step == :move_token_due_to_ambigious_city
+            # There should only be one token in the city
+            token = @city.tile.cities.map(&:tokens).find(&:any?).compact[0]
```
Comment (tobymao) [inline assets/app/view/game/part/city_slot.rb:57]: @city.tile.cities.flat_map(&:tokens).find(&:itself)

File: lib/engine/game/base.rb [hunk 3]
```diff
@@ -577,12 +580,22 @@ def init_tiles
         self.class::TILES.flat_map do |name, val|
           if val.is_a?(Integer)
             count = val
-            count.times.map { |i| Tile.for(name, index: i) }
+            count.times.map do |i|
+              Tile.for(name,
```
Comment (tobymao) [inline lib/engine/game/base.rb:584]: when i say each new line i mean

Tile.for(
name,
x,
y,
)

File: lib/engine/game/base.rb [hunk 4]
```diff
@@ -577,12 +580,22 @@ def init_tiles
...
+                       index: i,
+                       tile_reservation_blocks_others: self.class::TILE_RESERVATION_BLOCKS_OTHERS)
```
Comment (tobymao) [inline lib/engine/game/base.rb:586]: do tile reservations only block for certain places, or they block every single one?

Comment (jenf) [reply lib/engine/game/base.rb:586]: In 1830 they block every slot on the tile until the token is played

File: assets/app/view/game/hex.rb (hunk 1 above)
Comment (jenf) [reply assets/app/view/game/hex.rb:47]: Reposition sounds better, but I'll leave the action as move (1882 has a move private)

File: lib/engine/round/operating.rb [hunk 5]
```diff
@@ -300,12 +311,24 @@ def process_lay_tile(action)
           @teleported = ability[:hexes].include?(hex_id) &&
           ability[:tiles].include?(action.tile.name)
         end
+
+        new_tile = action.hex.tile
```
Comment (tobymao) [inline lib/engine/round/operating.rb:315]: this is confusing, just do action.tile

File: lib/engine/round/operating.rb [hunk 6]
```diff
@@ -300,12 +311,24 @@ def process_lay_tile(action)
...
+        if previous_tile.paths.empty? &&
+          new_tile.paths.any? &&
+          new_tile.cities.length > 1 &&
```
Comment (tobymao) [inline lib/engine/round/operating.rb:318]: always use size

File: lib/engine/round/operating.rb [hunk 7]
```diff
@@ -300,12 +311,24 @@ def process_lay_tile(action)
...
+          new_tile.cities.any? { |x| x.tokens.any? }
```
Comment (tobymao) [inline lib/engine/round/operating.rb:319]: new_tile.cities.flat_map(&:tokens).any?

File: lib/engine/round/operating.rb (hunk 5 above)
Comment (tobymao) [reply lib/engine/round/operating.rb:315]: also i don't think you have to dig into the tile, hex has paths as well

File: lib/engine/tile.rb [hunk 8]
```diff
@@ -102,7 +102,14 @@ def self.part(type, params, cache)
     end
 
     # rotation 0-5
-    def initialize(name, color:, parts:, rotation: 0, preprinted: false, index: 0, location_name: nil)
+    def initialize(name,
+                   color:,
+                   parts:,
```
Comment (tobymao) [inline lib/engine/tile.rb:107]: time to switch to **opts

File: lib/engine/tile.rb [hunk 9]
```diff
@@ -126,6 +133,7 @@ def initialize(name, color:, parts:, rotation: 0, preprinted: false, index: 0, l
       @preprinted = preprinted
       @index = index
       @blocks_lay = nil
+      @tile_reservation_blocks_others = tile_reservation_blocks_others
```
Comment (tobymao) [inline lib/engine/tile.rb:136]: rename to reservation_blocks or blocks_reservation

File: lib/engine/token.rb [hunk 10]
```diff
@@ -21,5 +21,11 @@ def swap!(other_token)
 
       @city.place_token(corporation, other_token)
     end
+
+    def move!(new_city)
+      @city.tokens.map! { |t| t == self ? nil : t }
```
Comment (tobymao) [inline lib/engine/token.rb:26]: i've seen this done a couple times, can we add a method that does this instead

File: lib/engine/round/operating.rb (hunk 5 above)
Comment (jenf) [reply lib/engine/round/operating.rb:315]: but doesn't have cities, but i've fixed the first comment.

Comment (tobymao) [reply lib/engine/round/operating.rb:315]: what do you mean?

File: assets/app/view/game/part/city_slot.rb [hunk 11]
```diff
@@ -52,7 +52,19 @@ def on_click(event)
 
           # If there's a choice of tokens of different types show the selector, otherwise just place
           next_tokens = @game.current_entity.tokens_by_type
-          if next_tokens.size == 1 || @game.round.step == :home_token
+          if @game.round.step == :reposition_token
+            # There should only be one token in the city
+            token = @city.tile.cities.flat_map(&:tokens).find(&:itself)
+            puts token
```
Comment (tobymao) [inline assets/app/view/game/part/city_slot.rb:58]: delete the dubginng code

File: lib/engine/round/operating.rb (hunk 5 above)
Comment (jenf) [reply lib/engine/round/operating.rb:315]: i'd need to dig into the tile anyway since i need to check if there's more than one city.

**Pull request merged**